addons/stock/models/stock_move_line.py

```diff
@@ -142,22 +143,41 @@ def _onchange_serial_number(self):
             message = None
             if self.lot_name or self.lot_id:
                 move_lines_to_check = self._get_similar_move_lines() - self
+                domain = [('product_id', '=', self.product_id.id),
+                          ('quantity', '!=', 0),
+                          '|', ('location_id.usage', '=', 'customer'),
+                               '&', ('company_id', '=', self.company_id.id),
+                                    ('location_id.usage', 'in', ('internal', 'transit'))]
                 if self.lot_name:
                     counter = Counter([line.lot_name for line in move_lines_to_check])
                     if counter.get(self.lot_name) and counter[self.lot_name] > 1:
                         message = _('You cannot use the same serial number twice. Please correct the serial numbers encoded.')
                     elif not self.lot_id:
-                        counter = self.env['stock.production.lot'].search_count([
-                            ('company_id', '=', self.company_id.id),
-                            ('product_id', '=', self.product_id.id),
-                            ('name', '=', self.lot_name),
-                        ])
-                        if counter > 0:
-                            message = _('Existing Serial number (%s). Please correct the serial number encoded.') % self.lot_name
+                        quants = self.env['stock.quant'].search([('lot_id.name', '=', self.lot_name)] + domain)
```

Comment:
We explicitly searched for lots in the move line company before the patch because the name could be the same across different companies. Here you only search for lot name. I think it could be an issue if you have 2 companies selling the same lot name

Reply:
By default, there is only one shared customer location. All the quants of delivery transfers of every company are located in the same location

Reply:
Ok, I can change the domain to not include customer locations and then problem solved? (The same problem will occur if I do the search on production.lot instead). It will just be confusing if the same company reuses a previously sold SN and then tries to sell it again. There won't be a warning during the receipt when reusing of a lot name, but there will be a validation error without a location context when using the SN in a delivery (this is what currently happens via `stock_quant.check_quantity`.)

I think it's clearer to give the warning when a SN is reused across all companies if it has been previously delivered since the user can choose to ignore the warning anyway rather than have no warning and have a more confusing validation error later on.

Reply:
Ah dang it, nevermind. I didn't realize the company_id for production_lot stays the same even after it's sold to the customer. Will do the search on that instead.

Reply:
Yep, I would just add company_id in the domain after `lot_name`

Reply:
Done. Got rid of the `domain` variable since it's no longer necessary. I had to add a filter to get the appropriate quants for the location info + make sure there's no warning in the case the SN already exists but its quants have been manually updated such that the SN name can be re-used.